pkg/controller/dns/external_dns_reconciler.go

```diff
@@ -20,19 +23,51 @@ func newExternalDnsReconciler(manager ctrl.Manager, resources []client.Object) e
 
 // NewExternalDns starts all resources required for external dns
 func NewExternalDns(manager ctrl.Manager, conf *config.Config, self *appsv1.Deployment) error {
-	if conf.DNSZoneDomain == "" {
+	if conf.PublicZoneConfig != nil && len(conf.PublicZoneConfig.ZoneIds) == 0 && conf.PrivateZoneConfig != nil && len(conf.PrivateZoneConfig.ZoneIds) == 0 {
 		return nil
 	}
 
-	dnsConfig := &manifests.ExternalDnsConfig{
-		ResourceName:  "external-dns",
-		TenantId:      conf.TenantID,
-		Subscription:  conf.DNSZoneSub,
-		ResourceGroup: conf.DNSZoneRG,
-		Domain:        conf.DNSZoneDomain,
-		RecordId:      conf.DNSRecordID,
-		IsPrivate:     conf.DNSZonePrivate,
-	}
-	objs := manifests.ExternalDnsResources(conf, self, dnsConfig)
+	configs, _ := generateZoneConfigs(conf)
+
+	// TODO: Uncomment this to implement an externalDNS cleanup runner
+	//err := newCleanupRunner(manager, namesToDelete)
+	//
+	//if err != nil {
+	//	return fmt.Errorf("failed to start cleanup runner: %w", err)
+	//}
+
+	objs := append([]client.Object{}, manifests.ExternalDnsResources(conf, self, configs)...)
+
 	return newExternalDnsReconciler(manager, objs)
 }
+
+func generateZoneConfigs(conf *config.Config) (configs []*manifests.ExternalDnsConfig, namesToDelete []string) {
+	publicResourceName := fmt.Sprintf("%s%s", manifests.ExternalDnsResourceName, manifests.PublicSuffix)
+	privateResourceName := fmt.Sprintf("%s%s", manifests.ExternalDnsResourceName, manifests.PrivateSuffix)
+
+	if conf.PrivateZoneConfig != nil && len(conf.PrivateZoneConfig.ZoneIds) > 0 {
+		configs = append(configs, generateConfig(conf, conf.PrivateZoneConfig, manifests.PrivateProvider, privateResourceName))
+	} else {
+		namesToDelete = append(namesToDelete, privateResourceName)
+	}
+
+	if conf.PublicZoneConfig != nil && len(conf.PublicZoneConfig.ZoneIds) > 0 {
+		configs = append(configs, generateConfig(conf, conf.PublicZoneConfig, manifests.PublicProvider, publicResourceName))
+	} else {
+		namesToDelete = append(namesToDelete, publicResourceName)
+	}
+
+	// namesToDelete will eventually be used for externalDNS cleanup
+	return
+}
+
+func generateConfig(conf *config.Config, dnsZoneConfig *config.DnsZoneConfig, provider manifests.Provider, resourceName string) *manifests.ExternalDnsConfig {
```

Comment:
Where is this used? On my phone so maybe I'm blind but I don't see this used anywhere

Reply:
used to generate externalDNS configs from the conf object. line 30 of this file and it's tested in the external_dns_reconciler test file.